tla/etcdraft.tla

```diff
@@ -0,0 +1,931 @@
+leaderVars == <<matchIndex, pendingConfChangeIndex>>
+
+\* @type: Int -> [jointConfig: Seq(Set(int)), learners: Set(int)]
+VARIABLE 
+    config
+VARIABLE 
+    reconfigCount
+
+configVars == <<config, reconfigCount>>
+
+VARIABLE 
+    durableState
+
+\* End of per server variables.
+----
+
+\* All variables; used for stuttering (asserting state hasn't changed).
+vars == <<messageVars, serverVars, candidateVars, leaderVars, logVars, configVars, durableState>>
+
+
+----
+\* Helpers
+
+\* The set of all quorums. This just calculates simple majorities, but the only
+\* important property is that every quorum overlaps with every other.
+Quorum(c) == {i \in SUBSET(c) : Cardinality(i) * 2 > Cardinality(c)}
+
+\* The set of all quorums. This just calculates simple majorities, but the only
+\* important property is that every quorum overlaps with every other.
+Quorums ==
+    [ s \in SUBSET Server |-> {i \in SUBSET(s) : Cardinality(i) * 2 > Cardinality(s)} ]
+
+\* The term of the last entry in a log, or 0 if the log is empty.
+\* @type: LOGT => Int;
+LastTerm(xlog) == IF xlog = <<>> THEN 0 ELSE xlog[Len(xlog)].term
+
+\* Helper for Send and Reply. Given a message m and bag of messages, return a
+\* new bag of messages with one more m in it.
+\* @type: (MSG, MSG -> Int) => MSG -> Int;
+WithMessage(m, msgs) == msgs (+) SetToBag({m})
+
+\* Helper for Discard and Reply. Given a message m and bag of messages, return
+\* a new bag of messages with one less m in it.
+\* @type: (MSG, MSG -> Int) => MSG -> Int;
+WithoutMessage(m, msgs) == msgs (-) SetToBag({m})
+
+\* Add a message to the bag of pendingMessages.
+SendDirect(m) == 
+    pendingMessages' = WithMessage(m, pendingMessages)
+
+\* All pending messages sent from node i
+PendingMessages(i) ==
+    FoldBag(LAMBDA x, y: IF y.msource = i THEN BagAdd(x,y) ELSE x, EmptyBag, pendingMessages)
+
+\* Remove all messages in pendingMessages that were sent from node i
+ClearPendingMessages(i) ==
+    pendingMessages (-) PendingMessages(i)
+
+\* Move all messages which was sent from node i in pendingMessages to messages
+SendPendingMessages(i) ==
+    LET msgs == PendingMessages(i)
+    IN /\ messages' = msgs (+) messages
+       /\ pendingMessages' = pendingMessages (-) msgs
+
+\* Remove a message from the bag of messages. Used when a server is done
+DiscardDirect(m) ==
+    messages' = WithoutMessage(m, messages)
+
+\* Combination of Send and Discard
+ReplyDirect(response, request) ==
+    /\ pendingMessages' = WithMessage(response, pendingMessages)
+    /\ messages' = WithoutMessage(request, messages)
+
+\* Default: change when needed
+ Send(m) == SendDirect(m)
+ Reply(response, request) == ReplyDirect(response, request) 
+ Discard(m) == DiscardDirect(m)
+     
+MaxOrZero(s) == IF s = {} THEN 0 ELSE Max(s)
+
+GetJointConfig(i) == 
+    config[i].jointConfig
+
+GetConfig(i) == 
+    GetJointConfig(i)[1]
+
+GetOutgoingConfig(i) ==
+    GetJointConfig(i)[2]
+
+IsJointConfig(i) ==
+    /\ GetJointConfig(i)[2] # {}
```

Comment:
Is `/\` needed here?

Reply:
It is ok for a single statement with or without `/\` before it. Both are correct and behave same.